File: GA/GA_Facility_Location_240326/problem.py
from __future__ import annotations

from pathlib import Path


class Problem:
    n = 0
    m = 0
    f: list[int] = []
    t: list[list[int]] = []

    @classmethod
    def read_instance(cls, file_name: str) -> None:
        input_file = Path(file_name)
        with input_file.open("r", encoding="utf-8") as reader:
            reader.readline()
            second_line = reader.readline().strip().split()
            rows = int(second_line[0])
            cols = int(second_line[1])

            cls.n = rows
            cls.m = cols
            cls.f = [0] * rows
            cls.t = [[0] * cols for _ in range(rows)]

            for y in range(rows):
                parts = reader.readline().strip().split()
                cls.f[y] = int(parts[1])
                for x in range(cols):
                    cls.t[y][x] = int(parts[2 + x])

    @classmethod
    def print_cost(cls) -> None:
        print("Transportkosten:")
        for row in cls.t:
            print(" ".join(str(value) for value in row))

        print("Fixkosten:")
        for value in cls.f:
            print(value)
        print("-------------------------------")

    @classmethod
    def fitness(cls, gene: list[int]) -> int:
        fit = 0

        for s, bit in enumerate(gene):
            if bit == 1:
                fit += cls.f[s]

        for customer in range(cls.m):
            min_cost = 2**31 - 1
            for facility, bit in enumerate(gene):
                if bit == 1 and cls.t[facility][customer] < min_cost:
                    min_cost = cls.t[facility][customer]
            fit += min_cost

        return fit

File: GA/GA_Facility_Location_240326/Problem.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;
//Optimal Value for 123UnifS.txt: 71342



public class Problem{
	
	// problembezogene Daten
	static int n;							//Number of Facilities
	static int m;							//Number of Customers
	static int[]   f;   					//Fix-Cost; cost of Facilities
	static int[][] t;						//Transportcost

	// verfahrensbezogene Daten
	//static double pm; 						//Mutationswahrscheinlichkeit

	
	public static void readInstance(String dateiName){
		// von Florian Haefner

		// Aufbau der Datei:
		// Dateiname
		// Zeilen   --- Spalten --- 0 (= Es gibt keine Kapazität)
		// Standort --- Fixkosten --- TransportkostenZuKunde1 --- ...
		// Standort --- ...
		// ...

		// Die Eingabedatei muss im Projektordner liegen!
		//File inputfile = new File("123UnifS.txt");
		File inputfile = new File(dateiName);
		
			
		FileReader filereader = null;
		BufferedReader reader = null;
		Scanner lineScanner   = null;

		try
		{
			filereader = new FileReader(inputfile);
			reader     = new BufferedReader(filereader);

			// Erste Zeile wird gelesen und direkt verworfen, da dort nur der Dateiname
			// steht.
			reader.readLine();

			String secondline = reader.readLine();
			lineScanner = new Scanner(secondline);
			lineScanner.useDelimiter(" ");
			int zeilen  = lineScanner.nextInt();//Anzahl Standorte
			int spalten = lineScanner.nextInt();//Anzahl Kunden
			n    = zeilen;
			m    = spalten;
			f    = new int[zeilen];
			t    = new int[zeilen][spalten];
			
			//System.out.println(zeilen + " " + spalten);
			
			for (int y = 0; y < zeilen; y++){
				lineScanner = new Scanner(reader.readLine());
				lineScanner.useDelimiter(" ");
				lineScanner.nextInt(); // Standortnummer
				f[y] = lineScanner.nextInt(); // Fixkosten
				
				for (int x = 0; x < spalten; x++)
				{
					t[y][x] = lineScanner.nextInt();
				}
			}
		}

		catch (IOException e)
		{
			e.printStackTrace();
		}

		finally
		{
			try
			{
				filereader.close();
				reader.close();
				lineScanner.close();
			}

			catch (IOException e)
			{
				e.printStackTrace();
			}
		}
	}
	
	
	
	
	public static void printCost() {
		System.out.println("Transportkosten:");		
		for(int i=0;i<t.length;i++) {
			for(int j=0;j<t[i].length;j++) {
				System.out.print(t[i][j] + " "); 
			}
			System.out.println();
		}
		
		System.out.println("Fixkosten:");
		for(int i=0;i<t.length;i++) {
			System.out.println(f[i]); 
		}
		System.out.println("-------------------------------");
		
	}		
	
	
	public static int fitness(int[] gene) {
		// gene: 110101
		
		int fit = 0;

		for (int s = 0; s < gene.length; s++) {
			if (gene[s] == 1) {
				fit += f[s];
			}
		}

		for (int k = 0; k < m; k++) {
			int min = Integer.MAX_VALUE;
			for (int s = 0; s < gene.length; s++) {
				if (gene[s] == 1) {
					if (t[s][k] < min) {
						min = t[s][k];
					}
				}
			}
			fit += min;
		}

		return fit;
	}
}
